include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/EdgeFunctions.h

```diff
@@ -472,6 +482,149 @@ template <typename AnalysisDomainTy> class EdgeFunctions {
   getSummaryEdgeFunction(n_t Curr, d_t CurrNode, n_t Succ, d_t SuccNode) = 0;
 };
 
+// This class can be used as a singleton factory for EdgeFunctions that only
+// have constant data.
+//
+// EdgeFunction that have constant data only need to be allocated once and can
+// be turned into a singleton, which save large amounts of memory. The
+// automatic singleton creation can be enbaled for a EdgeFunction class by
+// inheriting from EdgeFunctionSingletonFactory and only allocating
+// EdgeFunction throught the provided createEdgeFunction method.
+//
+// Old unused EdgeFunction that were deallocated need to/should be clean from
+// the factory by either calling cleanExpiredEdgeFunctions or initializing the
+// automatic cleaner thread with initEdgeFunctionCleaner.
+template <typename EdgeFunctionType, typename CtorArgT>
+class EdgeFunctionSingletonFactory {
+public:
+  EdgeFunctionSingletonFactory() = default;
+  EdgeFunctionSingletonFactory(const EdgeFunctionSingletonFactory &) = default;
+  EdgeFunctionSingletonFactory &
+  operator=(const EdgeFunctionSingletonFactory &) = default;
+  EdgeFunctionSingletonFactory(EdgeFunctionSingletonFactory &&) noexcept =
+      default;
+  EdgeFunctionSingletonFactory &
+  operator=(EdgeFunctionSingletonFactory &&) noexcept = default;
+
+  virtual ~EdgeFunctionSingletonFactory() {
+    std::lock_guard<std::mutex> DataLock(getCacheData().DataMutex);
+  }
+
+  // Creates a new EdgeFunction of type EdgeFunctionType, reusing the previous
+  // allocation if an EdgeFunction with the same values was already created.
+  static inline std::shared_ptr<EdgeFunctionType>
```

Comment:
shared_ptr again?

Reply:
As long as we have no other memory management in phasar we need to stick to the shared_ptr approach. If we get the memory manager working, the factory can be updated to work with the mem manager.